src/main/java/com/adobe/cq/commerce/graphql/client/impl/GraphqlClientAdapterFactory.java

```diff
@@ -44,31 +49,33 @@
 public class GraphqlClientAdapterFactory implements AdapterFactory {
 
     protected static final String RESOURCE_CLASS_NAME = "org.apache.sling.api.resource.Resource";
-
     protected static final String GRAPHQLCLIENT_CLASS_NAME = "com.adobe.cq.commerce.graphql.client.GraphqlClient";
-
     protected static final String CONFIG_NAME = "cloudconfigs/commerce";
 
     private static final Logger LOGGER = LoggerFactory.getLogger(GraphqlClientAdapterFactory.class);
 
-    protected Map<String, GraphqlClient> clients = new ConcurrentHashMap<>();
+    protected NavigableSet<Holder> clients = new ConcurrentSkipListSet<>();
+    protected Map<String, GraphqlClient> clientsByIdentifier = new ConcurrentHashMap<>();
+    protected BundleContext bundleContext;
 
     @Reference(
         service = GraphqlClient.class,
         bind = "bindGraphqlClient",
         unbind = "unbindGraphqlClient",
         cardinality = ReferenceCardinality.AT_LEAST_ONE,
         policy = ReferencePolicy.DYNAMIC)
-    protected void bindGraphqlClient(GraphqlClient graphqlClient, Map<?, ?> properties) {
+    protected void bindGraphqlClient(GraphqlClient graphqlClient, Map<String, Object> properties) {
         String identifier = graphqlClient.getIdentifier();
         LOGGER.info("Registering GraphqlClient '{}'", identifier);
-        clients.put(identifier, graphqlClient);
+        clients.add(new Holder(graphqlClient, properties));
+        clientsByIdentifier.remove(identifier);
```

Comment:
Remove and not put?

Reply:
The idea was to update the ordered set of graphql clients when ever a client binds/unbinds. The map of clients by identifier however was to be calculated on demand whenever a particular client was to be returned by getAdapter(). The remove cleared this calculation. This had the advantage that we did not require a multi-map like data structure to keep the references per identifier. 

Anyway, I see the point that this is not obvious and rebuilding the effective data structure that is used by getAdapter() here makes sense and reduces complexity . Also if we do so we can use a read only / non-concurrent data structure. Done in https://github.com/adobe/commerce-cif-graphql-client/pull/38/commits/8c637d132efe27c9c9e88f5ab0d13686bf3037e4